File: 2022/04.js
/**
 * Share your thoughts
 */

const year = '2022';
const day = '04';
const fs = require('fs');

const parseInput = () => {
    let filename = `./data/${day}`;
    if (process.argv.length > 2) {
        filename += '.test';
    }
    let data = fs.readFileSync(filename, 'utf8');
    return data;
}

const parsePairs = (data) => {
  return data.split('\n')
    .map(r => r.trim().split(',')
      .map(p => p.trim().split('-')
        .map(x => parseInt(x)
        )
      )
    ).filter(r => r.length === 2);
}

const checkOverlap = (p1, p2) => {
  if ((p1[0] <= p2[0] && p1[1] >= p2[1])
    || (p2[0] <= p1[0] && p2[1] >= p1[1])) {
    return true;
  }
  return false;
}

const checkPartialOverlap = (p1, p2) => {
  for (let i = p1[0]; i <= p1[1]; i++) {
    if (i >= p2[0] && i <= p2[1]) {
      return true;
    }
  }
  for (let i = p2[0]; i <= p2[1]; i++) {
    if (i >= p1[0] && i <= p1[1]) {
      return true;
    }
  }
  return false;
}

const part1 = (data) => {
  console.time('Time');
  let res = 0;

  const pairs = parsePairs(data);
  for (const pair of pairs) {
    if (checkOverlap(pair[0], pair[1])) {
      res += 1;
    }
  }

  console.log('Part 1:', res);
  console.timeEnd('Time');
}

const part2 = (data) => {
  console.time('Time');
  let res = 0;

  const pairs = parsePairs(data);
  console.log(pairs);
  for (const pair of pairs) {
    if (checkPartialOverlap(pair[0], pair[1])) {
      console.log('overlap', pair);
      res += 1;
    }
  }

  console.log('Part 2:', res);
  console.timeEnd('Time');
}

/**
 * Main
 */
console.log( `Advent Of Code ${year}-${day}`);
console.log('==========================');
part1(parseInput());
console.log('--------------------------');
part2(parseInput());

File: 2022/04.rs

/**
 * 2022 Day 04: Cleanning
 */

mod common;

use std::env;
use common::args::check_test_mode;
use common::data::parse_file;

fn main() {
    let year = 2022;
    let day = 4;

    let args: Vec<String> = env::args().collect();
    println!("Advent of Code {year} - Day {day:02}");

    let data: Vec<String> = parse_file(
        String::from(format!("./data/{day:02}")),
        check_test_mode(args),
    );

    println!("Part 1: {}", part_1(data.clone()));
    // println!("Part 2: {}", part_2(data.clone()));
}

fn part_1(data: Vec<String>) -> i32 {
    let mut res = 0;

    println!("{:?}", data);
    for pairs in data.iter() {
        let pair: Vec<String> = pairs
            .split(',')
            .map(|p| p.split('-').collect())
            .collect();
        println!("{} {}", pair[0][0], pair[0][1]);
    }

    return res;
}

